src/core/pile.sol

```diff
@@ -17,150 +17,203 @@ pragma solidity >=0.4.24;
+    }
+
+    function debt(uint loan) public view returns (uint) {
+        uint rate = group[loan];
+        uint chi = rates[rate].chi;
+        if (now >= rates[rate].rho) {
+            (chi,) = compound(rate);
+        }
+        return fromPie(chi, pie[loan]);
+    }
+
+    function rateDebt(uint rate) public view returns (uint) {
+        uint chi = rates[rate].chi;
+        uint pie = rates[rate].pie;
+        
+        if (now >= rates[rate].rho) {
+            (chi,) = compound(rate);
+        } 
+        return fromPie(chi, pie);
+    } 
+
+    // --- Interest Rate Group Implementation ---
+
+    // set rate group for a loan
+    function setRate(uint loan, uint rate) public auth {
+        require(pie[loan] == 0, "non-zero-debt");
+        group[loan] = rate;
+    }
+
+    // change rate group for a loan
+    function changeRate(uint loan, uint newRate) public auth {
+        uint currentRate = group[loan];
+        drip(currentRate);
+        drip(newRate);
+        uint pie_ = pie[loan];
+        uint debt = fromPie(rates[currentRate].chi, pie_);
+        rates[currentRate].pie = sub(rates[currentRate].pie, pie_);
+        pie[loan] = toPie(rates[newRate].chi, debt);
+        rates[newRate].pie = add(rates[newRate].pie, pie[loan]);
+        group[loan] = newRate;
+    }
+
+    // set/change the interest rate of a rate category
+    function file(uint rate, uint speed_) public auth {
+        require(speed_ != 0);
+        rates[rate].speed = speed_;
+        rates[rate].chi = ONE;
+        rates[rate].rho = uint48(now);
+        drip(rate);
+    }
+
+    // accrue neeeds to be called before any debt amounts are modified by an external component
+    function accrue(uint loan) public {
+        drip(group[loan]);
+    }
+    
+    // compound calculates the new chi, the delta between old and new chi and the delta of debt
+    function compound(uint rate) internal view returns (uint chi, uint delta) {
+        uint48 rho = rates[rate].rho;
+        uint speed = rates[rate].speed;
+        uint chi = rates[rate].chi;
+        uint pie = rates[rate].pie;
+        uint debt = fromPie(chi, pie);
+
+        // compounding in seconds
+        uint chi_ = rmul(rpow(speed, now - rho, ONE), chi);
+        require(chi != 0);
+        delta = fromPie(chi_, pie) - debt;
+
+        return (chi_, delta);
+    }
+
+    // drip updates the chi of the rate category by compounding the interest and
+    // updates the total debt
+    function drip(uint rate) public {
+        if (now >= rates[rate].rho) {
+            (uint chi, uint delta) = compound(rate);
+            rates[rate].chi = chi;
+            rates[rate].rho = uint48(now);
+            total = add(total, delta);
+        }
     }
 
-    function depend(bytes32 what, address data) public auth {
-        if (what == "lender") { lender = data; }
+    // convert debt amount to pie
+    function toPie(uint chi, uint pie) internal view returns (uint) {
+        return rdiv(pie, chi);
```

Comment:
We need to use the rounding div method here

Reply:
@manuelpolzhofer is that correct in tinlake-math? I'd strongly vote for refactoring this right now instead of reviewing this code here only to replace it a day or two later.